hazelcast/src/main/java/com/hazelcast/core/IMap.java

```diff
@@ -760,13 +760,99 @@
+     * <pre>{@code
+     *   // assuming an IMap<String, String>
+     *   ICompletableFuture<String> future = map.putAsync("a", "b", 5, TimeUnit.MINUTES);
+     *   future.andThen(new ExecutionCallback<String>() {
+     *     public void onResponse(String response) {
+     *       // do something with old value returned by put operation
+     *     }
+     *
+     *     public void onFailure(Throwable t) {
+     *       // handle failure
+     *     }
+     *   });
+     * }</pre>
+     * ExecutionException is never thrown.
+     * <p>
+     * <b>Warning 1:</b>
+     * <p>
+     * This method uses {@code hashCode} and {@code equals} of the binary form of
+     * the {@code key}, not the actual implementations of {@code hashCode} and {@code equals}
+     * defined in the {@code key}'s class.
+     * <p>
+     * <b>Warning 2:</b>
+     * <p>
+     * Time resolution for TTL is seconds. The given TTL value is rounded to the next closest second value.
+     * <p>
+     * <p><b>Note:</b>
+     * Use {@link #setAsync(Object, Object, long, TimeUnit)} if you don't need the return value, it's slightly
+     * more efficient.
+     *
+     * <p><b>Interactions with the map store</b>
+     * <p>
+     * If no value is found with {@code key} in memory,
+     * {@link MapLoader#load(Object)} is invoked to load the value from
+     * the map store backing the map. Exceptions thrown by load fail
+     * the operation and are propagated to the caller.
+     * <p>
+     * If write-through persistence mode is configured, before the value
+     * is stored in memory, {@link MapStore#store(Object, Object)} is
+     * called to write the value into the map store. Exceptions thrown
+     * by the store fail the operation and are propagated to the caller.
+     * <p>
+     * If write-behind persistence mode is configured with
+     * write-coalescing turned off,
+     * {@link com.hazelcast.map.ReachedMaxSizeException} may be thrown
+     * if the write-behind queue has reached its per-node maximum
+     * capacity.
+     *
+     * @param key         the key of the map entry
+     * @param value       the new value of the map entry
+     * @param ttl         maximum time for this entry to stay in the map (0 means infinite, negative means map config default)
+     * @param ttlUnit     time unit for the TTL
+     * @param maxIdle     maximum time for this entry to stay idle in the map.
+     *                    (0 means infinite, negative means map config default)
+     * @param maxIdleUnit time unit for the Max-Idle
      * @return ICompletableFuture from which the old value of the key can be retrieved
      * @throws NullPointerException if the specified key or value is null
      * @see ICompletableFuture
      * @see #setAsync(Object, Object, long, TimeUnit)
      */
-    ICompletableFuture<V> putAsync(K key, V value, long ttl, TimeUnit timeunit);
+    ICompletableFuture<V> putAsync(K key, V value, long ttl, TimeUnit ttlUnit, long maxIdle, TimeUnit maxIdleUnit);
```

Comment:
cant we have an api as in [ICache](https://github.com/hazelcast/hazelcast/blob/48769d33a7cf16900472b731ff77c744b3fc2118/hazelcast/src/main/java/com/hazelcast/cache/ICache.java#L250)? It seemeed easier to use to me.